### DISCUSSION TITLE
Tasmota ESP8266 Modbus+I2C+/OLED+1wire ?

### DISCUSSION THREAD

**Discussion by myozone** [arendst/Tasmota#11504] [Q&A] [answered]
Probably not the right place to ask - sorry if not.

Is it possible to use Tasmota with an energy meter (SDM120/230) and I2C pressure/temp/humidity (BME280) and a OLED display also with 1-Wire temperatures (DS18s20's) and a few IO's all at the same time ?? I'm guessing Tasmota - Sensors it would be possible ???

Thanks 
Dave.

**Comment by sfromis** [comment, marked as answer]
Yes, should be possible if you [compile a binary](https://tasmota.github.io/docs/Compile-your-build/) to include the mix of features, as the regular precompiled binaries do not combine I2C sensors and display. You should not try to add to -sensors or -display.

**Reply by myozone** [reply]
Ah ok, The sensors release would do it then without the display?. The OLED would be nice but really necessary

**Reply by barbudor** [reply]
[Smart-Meter](https://tasmota.github.io/docs/Smart-Meter-Interface/) is not included in any release so you will have to build your on custom firmware anyhow.
In that case, you can add sensors, display and all what you need in the same binary.
What Sfromis meant is that you should not try to add display to tasmota-sensors or sensors to tasmota-display.
But you can add all SML, sensors and displays to tasmota.

**Reply by myozone** [reply]
Ok I've got it now, I've just reflashed an ESP8266 to Sensors release and I can see SDMx20 listed, I hadn't realised it was an option like that, I assumed SDM's were a separate thing.

**Reply by sfromis** [reply]
The point was that while SDMx20 may be included in the sensors builld, the whole Smart Meter feature (which would probably be quite useful to you) is not included.

**Reply by myozone** [reply]
I've not looked into the Smart Meter Feature at all, all I need to do is push the energy and sensors data over MQTT to Node-red

**Reply by sfromis** [reply]
I've not at all used SDMxxx or the Smart Meter interface, but the text suggests that it may make sense for what you're after....

*This driver extracts selected values from Smart Meters over various protocols, filters and publishes them to MQTT as regular sensors.*

**Reply by myozone** [reply]
To be honest I don't have an SDM120 yet just figuring out how to use it when I get one. I first found [reaper7's](https://github.com/reaper7/SDM_Energy_Meter) repo then I thought Tasmota might do it with the extra sensors. Indeed the Smart Meter features look handy if I figure out what it/they actually work !

**Reply by myozone** [reply]
Does anyone know what the SDMx20 option/connection puts out ? display only, out over MQTT etc ??

**Reply by barbudor** [reply]
It lookslike SDM120 has quite a pretty native integration in Tasmota with a driver that extract all energy information.
So you may not need the Smart-Meter feature after all
If you want to have a look, this is driver `xnrg_08_sdm120.ino`
Field will be extracted and published in the SENSOR.ENERGY message at Teleperiod:
Voltage, current, active power, reactive power, apparent power, power factor, frequency, total active, import actiive (?), export active (?), import reactive (?), phase angle
I believe import/export are when you use this monitor a solar panel and waant to measure what you may return to the grid ?

To give you an idea about the SENSOR.ENERGY message, here is what a simple power measurement socket is publishing. It would be very similar with additional fields :
```
 tele/device-topic/SENSOR = {"Time":"2021-03-29T16:35:43+02:00","ENERGY":{"TotalStartTime":"2021-01-18T21:53:14","Total":0.437,"Yesterday":0.134,"Today":0.114,"Period":1,"Power":5,"ApparentPower":37,"ReactivePower":37,"Factor":0.13,"Voltage":306,"Current":0.121}}
```

**Reply by myozone** [reply]
Thanks @barbudor that's what I thought, I'm particularly interested in the VA (apparent power). Add in the 1-wire DS18s20's and the I2C BME280 and it's perfect !

Now to get an order for an SDM120 on a slow boat from China !

**Comment by Noschvie** [comment]
Does anyone know if the SDMx20 option works with a SDM230Modbus device ?
If not, I have to create a new driver based on `xnrg_08_sdm120.ino`, isn't it ?

**Reply by sfromis** [reply]
With a current docs example for [SDM230 (MODBus)](https://tasmota.github.io/docs/Smart-Meter-Interface/#sdm230-modbus), my guess is that the answer may be "yes".

**Reply by barbudor** [reply]
@Noschvie take care to use the right rS485 to TTL adapter
Tasmota do not support those which need a DIRection control signal, only those that can handle direction on their own

**Reply by Noschvie** [reply]
@barbudor have ordered this "RS485 Module" from https://www.hwhardsoft.de/english/webshop/accessories/ to use it with an ESP32. Is it the right one ?
(the SDM230Modbus device is ordered too...)
Thanks!

**Reply by barbudor** [reply]
Unfortunately that's the one to avoid :(
Example of a supported one:
https://www.aliexpress.com/item/1005001621746811.html
I know we need to make it clearer in the doc.  Adding that in my todo-list

**Reply by Noschvie** [reply]
ok, have to change my hardware setup...
Regarding the RS485 modules: is the difference "newer / older" or something else ?

**Reply by Noschvie** [reply]
> 
> 
> With a current docs example for [SDM230 (MODBus)](https://tasmota.github.io/docs/Smart-Meter-Interface/#sdm230-modbus), my guess is that the answer may be "yes".

sorry, not clear enough: if I have my hardware setup completed (next week..), which steps are needed afterwards?
No new driver needed, configuration as shown below should work?
![grafik](https://user-images.githubusercontent.com/48217146/114147918-7d477800-9919-11eb-9c7c-98709f68d057.png)
Thanks!

**Reply by barbudor** [reply]
Yes the above configuration should be fine although I always have a doubt if Tx/Rx are seen from the ESP or the meter point of view.

Regarding the module, RS485 is a bi-directionnal bus. The RS485 driver such as MAX485 has 4 pins towards the device side (and 2 pins towards the bus side).
1 pin is the reception (from bus to device)
1 pin is the transmission (from device to bus)
1 pin to enable the reception (enabled low)
1 pin to enable the transmission (enabled high)
So generally the 2 enable pins are tied together and connected to a GPIO on the device as a direction control.

Your adapter is just a MAX485 on a board and do not provide any logic related to direction.
For such bare adapter, the device software should leave the direction signal low to enable reception, and only when in need to talk to the bus, set the direction pin high which enable the device-to-bus (and disable reception to avoid local echo).

Tasmota do not implement the management of this direction pin but rely on a slightly more advanced adapter witch, on top of the MAX48, include some logic to drive the direction in "talking" mode as soon as the device is starting to transmit (and revert the direction into listening" mode when the transmission is completed.

**Reply by Noschvie** [reply]
Have completed the hardware setup as suggested, but don't get the SDM230 device connected.
Testing with ESP32 and D1mini running 9.3.1(sensors)
RS485 module connected to 3.3V, GND, TX, RX / A+, B- and GRD to SDM230
Is a terminating resistor needed at the SDM230 side?
How to debug ? Thanks!

**Reply by barbudor** [reply]
Looking at the code,
"SDMx20 TX" is the pin transmitting to the SDMx20, correctly configured on GPIO1 (ESP's TX), so it should be connected to the TX pin of your module
"SDMx20 RX" is the pin receiving from the SDMx20, correctly configured on GPIO3 (ESP's RX), so it should be connected to the RX pin of your module
You may need a resistor across A and B on the SDM side. Same value as the one on your module.

If the above is what you have wired, suggested reasons for failure could be:
- wrong baud rate : hardcoded to 2400 unless you override it with `#define SDM120_SPEED nnn`
- wrong modbus address : hardcoded to 1 unless you override it with `#define SDM120_ADDR n`
- SDM230 not compatible with "SDMx20" as implemented in Tasmota (the code explicitly refer to SDM120 and SDM220)

**Reply by Noschvie** [reply]
Got the RS485 / Modbus communication working, perfect, thanks!

**Comment by arendst** [comment]
You'll need this:

https://www.aliexpress.com/item/1005001621746811.html

and a config like 

![image](https://user-images.githubusercontent.com/11044339/114702539-79a25f80-9d24-11eb-827f-5c6f246e40d1.png)

Then connect A+ and A- to the SDM230-Modbus (Gnd is not needed) and Gnd, Vcc, Rx and Tx to the wemos.

Works for me.

NB. Make sure your SDM230 is the correct version with modbus support (SDM230-Modbus).
